crates/spirv-attrib/src/lib.rs

```diff
@@ -1,16 +1,34 @@
-#![cfg_attr(target_arch = "spirv", feature(register_attr), register_attr(spirv))]
+use core::iter::FromIterator;
+use core::iter::IntoIterator;
+use proc_macro::{Delimiter, Group, TokenStream, TokenTree};
 
-use proc_macro::TokenStream;
-use std::str::FromStr;
-
-#[cfg(not(target_arch = "spirv"))]
 #[proc_macro_attribute]
 pub fn spirv(_attr: TokenStream, item: TokenStream) -> TokenStream {
-    let mut item_str = item.to_string();
-
-    while let Some(start) = item_str.find("#[spirv(") {
-        let end = item_str[start..item_str.len()].find(")]");
-        item_str.replace_range(start..start + end.unwrap() + 2, "");
+    let mut tokens = Vec::new();
+    for tt in item {
+        if let TokenTree::Group(group) = tt.clone() {
+            if let Delimiter::Parenthesis = group.delimiter() {
+                let mut sub_tokens = Vec::new();
+                for tt in group.stream() {
+                    match tt.clone() {
+                        TokenTree::Group(group)
+                            if group.delimiter() == Delimiter::Bracket
+                                && matches!(group.stream().into_iter().next(), Some(TokenTree::Ident(ident)) if ident.to_string() == "spirv")
+                                && matches!(sub_tokens.last(), Some(TokenTree::Punct(p)) if p.as_char() == '#') =>
+                        {
+                            sub_tokens.pop();
+                        }
+                        _ => sub_tokens.push(tt),
+                    }
+                }
+                tokens.push(TokenTree::from(Group::new(
+                    Delimiter::Parenthesis,
+                    TokenStream::from_iter(sub_tokens),
+                )));
+                continue;
+            }
+        }
+        tokens.push(tt);
     }
-    TokenStream::from_str(item_str.as_str()).unwrap()
+    TokenStream::from_iter(tokens)
```

Comment:
Maybe this (and the other `from_iter` call) should/could be `tokens.into_iter().collect()`? I'm surprised `from_iter` can even take a `Vec`.

Reply:
`tokens.into_iter().collect()` Actually doesn't compile. it doesn't satisfy the FromIterator trait.